p2p/libp2p/discovery/hostDecorator.go

```diff
@@ -0,0 +1,182 @@
+package discovery
+
+import (
+	"context"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/libp2p/go-libp2p-core/connmgr"
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	protocol "github.com/libp2p/go-libp2p-core/protocol"
+	peerstore "github.com/libp2p/go-libp2p-peerstore"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+const (
+	hardCPSLimitLow  = 2
+	hardCPSLimitHigh = 100
+)
+
+type hostDecorator struct {
+	h            host.Host
+	acceptedChan chan struct{}
```

Comment:
This is over engineered and hard to understand (anonymous function and Connect function). Why not simplify this and pass an interface that has a method called `CanConnect(pi peer.AddrInfo)` ? The implementation of this interface can decide if it is ok to connect to that peer or not. 
cps parameter does not solve the whole problem of connecting to a peer just to disconnect from it milliseconds later (by the connectionMonitor implementation).

Reply:
That way we louse the ability to wait

Reply:
Disagree. We should avoid consuming node resources in a useless way.

Reply:
The scope of the PR is to introduce a way to limit the number of connections done by the discovery, And is doing just that.

Reply:
Is doing that in a non-deterministic and sub-optimal way.

Reply:
What exactly is non deterministic?